mpi4jax/_src/collective_ops/allreduce.py

```diff
@@ -164,13 +174,27 @@ def mpi_allreduce_xla_encode_gpu(ctx, x, token, op, comm, transpose):
         backend_config=descriptor,
     )
 
+    results = list(custom_call.results)
+    token = results[-1]
+    ctx.set_tokens_out(mlir.TokenSet({ordered_effect: (token,)}))
+
+    return results
+
 
 # This function evaluates only the shapes during AST construction
 def mpi_allreduce_abstract_eval(xs, token, op, comm, transpose):
-    return (
-        ShapedArray(xs.shape, xs.dtype),
-        core.abstract_token,
-    ), {effect}
+    if not transpose:
+        return (
+            ShapedArray(xs.shape, xs.dtype),
+            core.abstract_token,
+        ), {ordered_effect}
+    else:
+        # The transposition of an allreduce is just the identity, so it can be reordered
+        # and does not come with an ordered effect.
+        return (
+            ShapedArray(xs.shape, xs.dtype),
+            core.abstract_token,
+        ), {}
```

Comment:
@dionhaefner this is correct, right? (If I return `ordered_effect` here I would get an error in the translation rule because if `transpose is True` we don't set the input and output tokens, as we do nothing.

Reply:
... probably? I would be fine with setting effects either way to err on the side of caution, but your call.